pkg/tarbackfill/tarbackfill.go

```diff
@@ -0,0 +1,159 @@
+	currentIsQueued bool
+	err             error
+}
+
+// Backfiller is a wrapper for Backfill, which can supply headers to insert
+// into an archive which is on its way to being extracted.
+type Backfiller interface {
+	// Backfill either returns an entry for the passed-in path, nil if
+	// no entry should be added to the stream, or an error if something
+	// unexpected happened.
+	Backfill(string) (*tar.Header, error)
+}
+
+// NewReaderWithBackfiller creates a new Reader reading from r, asking the
+// passed-in Backfiller for information about parent directories which it
+// hasn't seen yet.
+func NewReaderWithBackfiller(r *tar.Reader, backfiller Backfiller) *Reader {
+	reader := &Reader{
+		Reader:     r,
+		backfiller: backfiller,
+		seen:       make(map[string]struct{}),
+	}
+	return reader
+}
+
+// Next returns either the next item from the archive we're filtering, or a
+// synthesized entry for a directory that arguably should have been in that
+// archive.
+func (r *Reader) Next() (*tar.Header, error) {
+	// Drain the queue first.
+	if len(r.queue) > 0 {
+		next, queue := r.queue[0], r.queue[1:]
+		r.queue = queue
+		r.currentIsQueued = len(r.queue) > 0
+		return next, nil
+	}
+	// If we've hit the end of the archive, we've hit the end of the archive.
+	r.currentIsQueued = false
+	if r.err != nil {
+		return nil, r.err
+	}
+	// Check what's next in the archive.
+	hdr, err := r.Reader.Next()
+	if err != nil {
+		r.err = err
+	}
+	if hdr == nil {
+		return hdr, err
+	}
+	for {
+		// Trim off an initial or final path separator.
+		name := strings.Trim(hdr.Name, "/")
+		if hdr.Typeflag == tar.TypeDir {
+			// Trim off an initial or final path separator, and
+			// note that we won't need to supply it later.
+			r.seen[name] = struct{}{}
+		}
+		// Figure out which directory this item is directly in.
+		p := name
+		dir, _ := path.Split(name)
+		var newHdr *tar.Header
+		for dir != p {
+			var bfErr error
+			dir = strings.Trim(dir, "/")
+			// If we already saw that directory, no need to interfere (further).
+			if _, ok := r.seen[dir]; dir == "" || ok || dir == name {
+				return hdr, err
+			}
+			// Ask the backfiller what to do here.
+			newHdr, bfErr = r.backfiller.Backfill(dir)
+			if bfErr != nil {
+				r.err = bfErr
+				return nil, bfErr
+			}
+			if newHdr == nil {
+				dir, _ = path.Split(dir)
+				continue
+			}
+			// Make sure the Name looks right, then queue up the current entry.
+			newHdr.Format = tar.FormatPAX
+			newHdr.Name = strings.Trim(newHdr.Name, "/")
+			if newHdr.Typeflag == tar.TypeDir {
+				// We won't need to supply it later.
+				r.seen[newHdr.Name] = struct{}{}
+				newHdr.Name += "/"
+			}
+			r.queue = append([]*tar.Header{hdr}, r.queue...)
+			hdr = newHdr
+			r.currentIsQueued = true
+			dir, _ = path.Split(dir)
+		}
+	}
+}
+
+// Read will either read from a real entry in the archive, or pretend very hard
+// that an entry we inserted had no content.
+func (r *Reader) Read(b []byte) (int, error) {
```

Comment:
What … happens … when the parent layer contains a regular file in place of an expected parent directory?

I suppose it doesn’t matter what we do here, because extracting the current layer should fail, shouldn’t it?